File: day1/part-2.js
const { readFileAndGetColumns } = require("./utils");

const calculateTotalSimilarity = ({ left = [], right = [] } = {}) => {
  let totalSimilarity = 0;

  for (let i = 0; i < left.length; i++) {
    const occurrencesInRight = right.filter((n) => n === left[i]);
    const similarity = left[i] * occurrencesInRight.length;

    totalSimilarity += similarity;
  }

  return totalSimilarity;
};

const executeExercise = async () => {
  const { left, right } = await readFileAndGetColumns();

  const totalSimilarity = calculateTotalSimilarity({ left, right });

  console.log({ totalSimilarity });
};

executeExercise();

File: day1/part-2.py
import numpy as np
import os

data_path = os.path.join(os.getcwd(), "data.txt")
data_np = np.loadtxt(data_path, dtype=(int, int))
left_np = data_np[:, 0]
rigth_np = data_np[:, 1]


def calculate_total_similarity(left, right):
    left_occurences_in_rigth = [
        np.count_nonzero(item == right) for item in left
    ]

    similarities = left * left_occurences_in_rigth

    return np.sum(similarities)


total_similarity = calculate_total_similarity(left=left_np, right=rigth_np)
print(total_similarity)
